### PR TITLE
[CDEC-342] Add Hedgehog generators for data types in infra

### PR DESCRIPTION
**Pull request by intricate**
## Description

Implement Hedgehog generators for serializable data types in infra.

## Linked issue

[https://iohk.myjetbrains.com/youtrack/issue/CDEC-342](https://iohk.myjetbrains.com/youtrack/issue/CDEC-342)
## Type of change

- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist

- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [ ] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [ ] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.

## Testing checklist

- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: infra/test/Test/Pos/Infra/Gen.hs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+import           Universum
+
+import           Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
+import           Data.Time.Units (Millisecond, Microsecond,
+                                  fromMicroseconds)
+
+import           Pos.Arbitrary.Core ()
```
Comment (ruhatch) [inline infra/test/Test/Pos/Infra/Gen.hs]: Why is this here?

File: infra/test/Test/Pos/Infra/Gen.hs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+genMillisecond :: Gen Millisecond
+genMillisecond =
+    fromMicroseconds <$> (toInteger <$> Gen.int Range.constantBounded)
+
+genMicrosecond :: Gen Microsecond
+genMicrosecond =
+    fromMicroseconds <$> (toInteger <$> Gen.int Range.constantBounded)
+
+-- TimeDiff is from core so this generator will be moved eventually.
```
Comment (intricate) [inline infra/test/Test/Pos/Infra/Gen.hs]: @ruhatch Do you think I should wait until we get the `core` generators done before merging this PR? Since these generators will be included with the `core` gens.

File: infra/test/Test/Pos/Infra/Gen.hs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+genInvMsg genA = InvMsg <$> genA
+
+genReqMsg :: Gen (Maybe a) -> Gen (ReqMsg a)
+genReqMsg genMA = ReqMsg <$> genMA
+
+genResMsg :: Gen a -> Gen (ResMsg a)
+genResMsg genA = ResMsg <$> genA <*> Gen.bool
+
+genMempoolMsg :: Gen (MempoolMsg a)
+genMempoolMsg = return MempoolMsg
```
Comment (ruhatch) [inline infra/test/Test/Pos/Infra/Gen.hs]: It's nice to use `pure` instead of `return`

File: infra/test/Test/Pos/Infra/Gen.hs [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+genDataMsg :: Gen a -> Gen (DataMsg a)
+genDataMsg genA = DataMsg <$> genA
+
+genDHTKey :: Gen DHTKey
+genDHTKey = do
+    b <- gen32Bytes
+    let k = bytesToDHTKey b :: Either String DHTKey
+    case k of
+        Left _   -> error "Failed to generate a DHTKey."
+        Right dk -> return dk
```
Comment (ruhatch) [inline infra/test/Test/Pos/Infra/Gen.hs]: And `pure` here

File: infra/test/Test/Pos/Infra/Gen.hs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+genDHTData :: Gen DHTData
+genDHTData = return $ DHTData ()
```
Comment (ruhatch) [inline infra/test/Test/Pos/Infra/Gen.hs]: And here :)

File: infra/test/Test/Pos/Infra/Gen.hs (hunk 2 above)
Comment (ruhatch) [reply infra/test/Test/Pos/Infra/Gen.hs]: Yes and they should really be moved to `Test.Pos.Util.Gen` I think - push them right to the bottom of the package stack!

File: infra/test/Test/Pos/Infra/Gen.hs [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+genDataMsg :: Gen a -> Gen (DataMsg a)
+genDataMsg genA = DataMsg <$> genA
+
+genDHTKey :: Gen DHTKey
+genDHTKey = do
+    b <- gen32Bytes
+    let k = bytesToDHTKey b :: Either String DHTKey
+    case k of
+        Left _   -> error "Failed to generate a DHTKey."
```
Comment (mhuesch) [inline infra/test/Test/Pos/Infra/Gen.hs]: Should we pass on the error string?

File: infra/test/Test/Pos/Infra/Gen.hs [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+genMempoolMsg :: Gen (MempoolMsg a)
+genMempoolMsg = return MempoolMsg
+
+genDataMsg :: Gen a -> Gen (DataMsg a)
+genDataMsg genA = DataMsg <$> genA
+
+genDHTKey :: Gen DHTKey
+genDHTKey = do
+    b <- gen32Bytes
```
Comment (mhuesch) [inline infra/test/Test/Pos/Infra/Gen.hs]: We could also inline `gen32Bytes` since it's only used once.

File: infra/test/cardano-sl-infra-test.cabal [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+                     , cardano-sl-infra
+                     , hedgehog
+                     , kademlia
+                     , memory
+                     , time-units
+                     , universum
+
+  default-language:    Haskell2010
+
+  default-extensions:  DeriveDataTypeable
```
Comment (intricate) [inline infra/test/cardano-sl-infra-test.cabal]: I need to take out the `default-extensions` as well and stick to specifying them at the module-level.

File: infra/test/cardano-sl-infra-test.cabal [hunk 9]
```diff
@@ -0,0 +1,33 @@
+name:                cardano-sl-infra-test
+version:             1.1.0
```
Comment (erikd) [inline infra/test/cardano-sl-infra-test.cabal]: I think that should actual be `1.3.0` to match the other packages.

Comment (intricate) [reply infra/test/cardano-sl-infra-test.cabal]: Do you mean `1.2.0`? Just clarifying.

Comment (intricate) [reply infra/test/cardano-sl-infra-test.cabal]: Sorry, @Jimbo4350 just pointed me to https://github.com/input-output-hk/cardano-sl/pull/3064

Comment (intricate) [reply infra/test/cardano-sl-infra-test.cabal]: So we're indeed going with `1.3.0`.

File: infra/test/Test/Pos/Infra/Gen.hs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+----------------------------------------------------------------------------
+
+genEpochSlottingData :: Gen EpochSlottingData
+genEpochSlottingData = EpochSlottingData <$> genMillisecond <*> genTimeDiff
+
+genSlottingData :: Gen SlottingData
+genSlottingData = createSlottingDataUnsafe <$> genMap
+  where
+    genMap :: Gen (Map EpochIndex EpochSlottingData)
+    genMap = Gen.map Range.constantBounded genEpochIndexDataPair
```
Comment (Jimbo4350) [inline infra/test/Test/Pos/Infra/Gen.hs]: `Range.constantBounded` here has the potential to hang your machine while running tests!

File: infra/test/Test/Pos/Infra/Gen.hs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    sd <- genEpochSlottingData
+    pure (i, sd)
+
+----------------------------------------------------------------------------
+-- Helper Generators
+----------------------------------------------------------------------------
+
+genMillisecond :: Gen Millisecond
+genMillisecond =
+    fromMicroseconds <$> (toInteger <$> Gen.int Range.constantBounded)
```
Comment (ruhatch) [inline infra/test/Test/Pos/Infra/Gen.hs]: This should be in `util`!

File: infra/test/Test/Pos/Infra/Gen.hs (hunk 10 above)
Comment (intricate) [reply infra/test/Test/Pos/Infra/Gen.hs]: Oops, my bad.

File: infra/test/Test/Pos/Infra/Gen.hs (hunk 11 above)
Comment (intricate) [reply infra/test/Test/Pos/Infra/Gen.hs]: Makes sense.
